crates/dpapi/src/crypto/mod.rs

```diff
@@ -174,35 +179,38 @@ pub fn content_encrypt(
 }
 
 pub fn kdf(algorithm: HashAlg, secret: &[u8], label: &[u8], context: &[u8], length: usize) -> CryptoResult<Vec<u8>> {
-    use rust_kbkdf::{kbkdf, CounterMode, InputType, KDFMode, SpecifiedInput};
-
-    let mut derived_key = vec![0; length];
-
-    macro_rules! kdf {
-        ($sha:ident) => {{
-            let key = HmacShaPrfKey::new(secret);
-            let mut hmac = $sha::new();
-
-            // KDF(HashAlg, KI, Label, Context, L)
-            // where KDF is SP800-108 in counter mode.
-            kbkdf(
-                &KDFMode::CounterMode(CounterMode { counter_length: 32 }),
-                &InputType::SpecifiedInput(SpecifiedInput { label, context }),
-                &key,
-                &mut hmac,
-                &mut derived_key,
-            )?;
-        }};
+    use hmac_pre::Hmac;
+    use kbkdf::{Counter, Kbkdf, Params};
+    use kout::*;
+
+    macro_rules! derive_key {
+        ($prf_ty:ty, { $($l_value:expr => $l_ty:ty,)* }) => {
+            $(
+                if $l_value == length {
+                    return Ok(Counter::<$prf_ty, $l_ty>::default()
+                        .derive(Params::builder(secret).with_label(label).with_context(context).use_counter(true).use_l(true).build())?
+                        .to_vec());
+                }
+            )*
+        };
     }
 
     match algorithm {
-        HashAlg::Sha1 => kdf!(HmacSha1Prf),
-        HashAlg::Sha256 => kdf!(HmacSha256Prf),
-        HashAlg::Sha384 => kdf!(HmacSha384Prf),
-        HashAlg::Sha512 => kdf!(HmacSha512Prf),
+        HashAlg::Sha1 => {
+            derive_key!(Hmac<sha1_pre::Sha1>, { 32 => Kout32, 64 => Kout64, });
+        }
+        HashAlg::Sha256 => {
+            derive_key!(Hmac<sha2_pre::Sha256>, { 32 => Kout32, 64 => Kout64, });
+        }
+        HashAlg::Sha384 => {
+            derive_key!(Hmac<sha2_pre::Sha384>, { 32 => Kout32, 64 => Kout64, });
+        }
+        HashAlg::Sha512 => {
+            derive_key!(Hmac<sha2_pre::Sha512>, { 32 => Kout32, 64 => Kout64, });
+        }
     }
```

Comment:
I can generate this `match` in the macro too, but that way macro will become too complex and unreadable (example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3d37bce0287bb2a063d51cf24c48c2ae)

Reply:
It’s fine to keep the match spelled out like this 👍 
It’s not like we are going to write it elsewhere.